ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -59,6 +59,30 @@ static const void *newsubmsghandlerdata(upb_handlers* h, uint32_t ofs,
   return hd;
 }
 
+typedef struct {
+  size_t ofs;      // union data slot
+  size_t case_ofs; // oneof_case field
+  uint32_t tag;    // tag number to place in data slot
+  const upb_msgdef *md; // msgdef, for oneof submessage handler
+} oneof_handlerdata_t;
+
+static const void *newoneofhandlerdata(upb_handlers *h,
+                                       uint32_t ofs,
+                                       uint32_t case_ofs,
+                                       const upb_fielddef *f) {
+  oneof_handlerdata_t *hd = ALLOC(oneof_handlerdata_t);
+  hd->ofs = ofs;
+  hd->case_ofs = case_ofs;
+  hd->tag = upb_fielddef_number(f);
```

Comment:
C++ does create a separate ID space -- field indices into the oneof -- but it also exposes `[oneof_name]_case()` to the user, so the IDs may matter a bit more. We only use this internally, and never iterate through the space, so any number that is unique per field will do, I think, and absent any other requirements, it seems simpler to just use the tag number that's there? (Field index would also do just as well.)